rclpy/src/rclpy/_rclpy.c

```diff
@@ -3369,6 +3371,382 @@ rclpy_clock_set_ros_time_override(PyObject * Py_UNUSED(self), PyObject * args)
+      }
+      PyList_SET_ITEM(value, i, member_value);
+    }
+  } else if (variant->double_array_value) {
+    type_enum_value = rcl_interfaces__msg__ParameterType__PARAMETER_DOUBLE_ARRAY;
+    value = PyList_New(variant->double_array_value->size);
+    if (NULL == value) {
+      return NULL;
+    }
+    for (size_t i = 0; i < variant->double_array_value->size; i++) {
+      member_value = PyFloat_FromDouble(variant->double_array_value->values[i]);
+      if (NULL == member_value) {
+        Py_DECREF(value);
+        return NULL;
+      }
+      PyList_SET_ITEM(value, i, member_value);
+    }
+  } else if (variant->string_array_value) {
+    type_enum_value = rcl_interfaces__msg__ParameterType__PARAMETER_STRING_ARRAY;
+    value = PyList_New(variant->string_array_value->size);
+    if (NULL == value) {
+      return NULL;
+    }
+    for (size_t i = 0; i < variant->string_array_value->size; i++) {
+      member_value = PyUnicode_FromString(variant->string_array_value->data[i]);
+      if (NULL == member_value) {
+        Py_DECREF(value);
+        return NULL;
+      }
+      PyList_SET_ITEM(value, i, member_value);
+    }
+  }
+
+  PyObject * args = Py_BuildValue("(i)", type_enum_value);
+  if (NULL == args) {
+    Py_DECREF(value);
+    return NULL;
+  }
+  PyObject * type = PyObject_CallObject(parameter_type_cls, args);
+  Py_DECREF(args);
+  args = Py_BuildValue("OOO", name, type, value);
+  if (NULL == args) {
+    Py_DECREF(type);
+    Py_DECREF(value);
+    return NULL;
+  }
+  Py_DECREF(value);
+  Py_DECREF(type);
+
+  PyObject * param = PyObject_CallObject(parameter_cls, args);
+  Py_DECREF(args);
+  return param;
+}
+
+/// Populate a Python dict with a dict of node parameters by node name
+/**
+ * On failure a Python exception is raised and false is returned
+ *
+ * \param[in] name The name of the parameter
+ * \param[in] variant The variant object to create a Parameter from
+ * \param[in] parameter_cls The PythonObject for the Parameter class.
+ * \param[in] parameter_type_cls The PythonObject for the Parameter.Type class.
+ * \param[out] node_params_dict The PythonObject to populate with node names and parameters.
+ *
+ * Returns true when parameters are set successfully
+ *         false when there was an error during parsing.
+ */
+static bool
+_populate_node_parameters_from_rcl_params(
+  const rcl_params_t * params, rcl_allocator_t allocator, PyObject * parameter_cls,
+  PyObject * parameter_type_cls, PyObject * node_params_dict)
+{
+  for (size_t i = 0; i < params->num_nodes; i++) {
+    PyObject * py_node_name;
+    if (params->node_names[i][0] != '/') {
+      py_node_name = PyUnicode_FromString(
+        rcutils_format_string(allocator, "/%s", params->node_names[i]));
+    } else {
+      py_node_name = PyUnicode_FromString(params->node_names[i]);
+    }
+    if (NULL == py_node_name) {
+      return false;
+    }
+    PyObject * parameter_dict;
+    if (!PyDict_Contains(node_params_dict, py_node_name)) {
```

Comment:
I did a bit of empirical testing and it can return -1 when the item you pass in as a dict isn't. Since we create the dict that doesn't seem like a case worth checking at runtime.